add bancor views

milkyklim opened the pull request:
I've checked that:

* [x] the folder name matches the schema name
* [x] the schema name exists in Dune
* [x] the view name starts with `view_`
* [x] the filename matches the defined view and ends with .sql
* [x] each file has only one view defined  
* [x] column names are `lowercase_snake_cased`

milkyklim commented:
@mewwts 
- Can you check that the views are alright?  
- Should I add missing tokens that are used by bancor to `erc20.tokens` as a separate PR?

hagaetc commented:
Hey @milkyklim - nice work as always:) 

I think it best if you join or union all the conversion events together - they are different tables in Dune because they have different ABIs for different versions, but it makes sense to query them as one table. 

Yes, please add missing tokens in `erc20.tokens` in a separate PR.

milkyklim commented:
Will update, thanks for review!

milkyklim commented:
@hagaetc I've converted multiple events into one [view_convert.sql](https://github.com/duneanalytics/abstractions/blob/513231e634ac94e7c98fa0835ae09c49f9a31688/schema/bancor/view_convert.sql)

However, I am not 100% sure that all events are actually in Dune DBs, e.g. for this [transaction](https://etherscan.io/tx/0x00001382ccbbd5fd6f5e66dbe7eb4c8405c447e2fe41e4b50e7ece6097997d52) includes 3 `Conversion`s and they are in DB, while this [transaction](https://etherscan.io/tx/0x00001a7af2bf99e5117d0403b8588bad6cc4f2f9298dafe6c6e805467aae89c0) also has 3 events but DB shows only 2.

mewwts commented:
FYI I'm just waiting for @hagaetc and the bancor team to confirm!

hagaetc commented:
Hey @milkyklim - sorry for the delay here. It is rather hard to get all the Bancor data but all recent conversions should be included in `bancor."BancorConverterWithFee_evt_Conversion"` - the missing event you mentioned is from 2 years ago. 

I am not familiar with the details of Bancor, but I assume the BNT<>USDB cases are solved correctly. Not that you can make a PR to `erc20.tokens` as well if we lack symbols for some tokens.

All over I think this looks good and can be merged :)

milkyklim commented:
@hagaetc, you seem to be right about `bancor."BancorConverterWithFee_evt_Conversion"` but I wanted to make views as full as possible, specifically:

- `bancor."BancorConverter_evt_Conversion"` is fully contained in `bancor."BancorConverterWithFee_evt_Conversion"`
- `bancor."BancorNetwork_v1_evt_Conversion"` partially overlaps with `bancor."BancorConverterWithFee_evt_Conversion"`
- `bancor."BancorLegacyConverter_evt_Conversion"` partially overlaps with `bancor."BancorConverterWithFee_evt_Conversion"`

Therefore, I've added this huge union in [view_convert.sql](https://github.com/duneanalytics/abstractions/blob/57a79031f10d241fef57988cc8d5ace8183bb95f/schema/bancor/view_convert.sql). 

`BNT<>USDB` exception is there cause `BNT<>USDB` pool won't be processed correctly. :) 

I've fixed the issue with the semicolons, see above.

hagaetc commented:
✅✅Sounds good!

milkyklim commented:
@mewwts need some help on your side. 

Can you double check that following tables and values are parsed correctly:
- bancor."BancorConverterWithFee_evt_Conversion", `trader` column is a contract address but should not it be equal to transaction sender?
- bancor."BancorLegacyConverter_evt_Conversion", `trader` same.
- bancor."BancorConverter_evt_Conversion", `trader` same.

mewwts commented:
Happy to help, can you link me some examples on telegram?

milkyklim commented:
@hagaetc these are the final changes from my side.

mewwts commented on schema/bancor/view_convert.sql:
```diff
@@ -0,0 +1,80 @@
+    AND p2.symbol = t2.symbol
+;
+
+CREATE INDEX IF NOT EXISTS view_convert_idx_1 ON bancor.view_convert (source_token_address) INCLUDE (source_token_amount);
+CREATE INDEX IF NOT EXISTS view_convert_idx_2 ON bancor.view_convert (target_token_address) INCLUDE (target_token_amount);
+CREATE INDEX IF NOT EXISTS view_convert_idx_3 ON bancor.view_convert (block_time);
```
Non-materialized views don't have indexes, so these lines should be removed.

mewwts commented on schema/bancor/view_smart_tokens.sql:
```diff
@@ -0,0 +1,274 @@
+    ('ZRXUSDB'::text, '\x1a3c6768e200482f5f47d1be77b7255abcae4fe2'::bytea, 18::numeric)
+;
+
+CREATE UNIQUE INDEX IF NOT EXISTS view_smart_tokens_id ON bancor.view_smart_tokens (contract_address);
+CREATE INDEX view_tokens_1 ON bancor.view_smart_tokens (symbol);
```
This view is non-materialized, so can't have indexes.

mewwts commented on schema/bancor/view_smart_tokens.sql:
```diff
@@ -0,0 +1,274 @@
+CREATE OR REPLACE VIEW bancor.view_smart_tokens (symbol, contract_address, decimals) AS VALUES
```
Are there no events when smart-tokens are added, such that we could compute this dynamically?

milkyklim replied:
These are created in `BancorConverterRegistry_evt_SmartTokenAdded`. Problem, event doesn't contain `symbol` and `decimals`. If you have an idea how to fix this – glad to help.

For ref:
```
SELECT *
FROM bancor."BancorConverterRegistry_evt_SmartTokenAdded";
```

mewwts replied:
Got it! Yeah, for now this is OK!

mewwts commented on schema/erc20/tokens.sql:
```diff
@@ -143,12 +181,46 @@ COPY erc20.tokens (contract_address, symbol, decimals) FROM stdin;
 \\x94a2aaa374a8f2d52dad24330c8a0ec2934700ae	FCOUSDB	18
 \\x009e864923b49263c7f10d19b7f8ab7a9a5aad33	FKX	18
 \\xf04a8ac553fcedb5ba99a64799155826c136b0be	FLIXX	18
+\\x01fb4a9a68dd46852addfa98321def44289b7d8d	FNCD	18
+\\x04b7c9b88c1d33f0293449aa81b143272bcb2a8b	FNCD	18
+\\x07be1d613ec9a7965bbe98064e66e3b3d648da66	FNCD	18
+\\x151a1eb5a1c989cdca7a48c07a1a5ae0ef80ea0d	FNCD	18
+\\x15887c0a2ee20e2b6f22a977b18bce47c5c5a2f2	FNCD	18
+\\x2561df6c7ee681864c8a1245aa51a06a1e22a52b	FNCD	18
+\\x26157a6e4f84c49942e32484d4728aa6e05c6417	FNCD	18
+\\x304b137223b0b240ba9e86d49aafc152a11b6dc2	FNCD	18
+\\x3796878de516db33cf1cba0b7d333cd1db96434e	FNCD	18
+\\x3798e0b7453b2954d61cbc86bbea2d52292531bf	FNCD	18
+\\x536cb3bafccb17537ba581b0ac4355f7e19df395	FNCD	18
+\\x6ec5a9d14d5206cad15b32e4bbe5e0b0be30ce14	FNCD	18
+\\x84a9986f4d04e73bc9442245b8dfef7f77305637	FNCD	18
+\\x8e0f05cd0e4e9dc28dbba761aac1f59337d1c455	FNCD	18
+\\x910822f6844304afb34ae8d7db619cdfa498ad49	FNCD	18
+\\x994e4008bcdc6cf08094fd08bb4418e70b6d686d	FNCD	18
+\\x9fd85ca57f20e7a39726e0e9cfad51f45bb2b0e3	FNCD	18
+\\xa0bbe61f8705d0dde7fa46f78a148c913b260c35	FNCD	18
+\\xb2db902091500cd04f41396d664dfa653517af68	FNCD	18
+\\xb7f5497f2ae792d59608c9a046b141cd29f3b526	FNCD	18
+\\xbfc615ab171d10a4e6498549cd3a46766feefe01	FNCD	18
+\\xc3308318eca1eab76d2f4b52c310289ae526c8ca	FNCD	18
+\\xce6437ee4932cfba3ecaefb93a1929a379b13b24	FNCD	18
+\\xcf97b4f961fa247cd60eb3cbf92a135fa2c9cc5d	FNCD	18
+\\xd2a3e62ca1658421a8a711bd1f8ed9d01780e6e6	FNCD	18
+\\xd609609fdf3a6ea89d5ea981af5ce5c3c3daa44a	FNCD	18
+\\xe23a3a78b131553c0bab4323bba7f61f6950bdaa	FNCD	18
+\\xf0fd2500bd31a56cf1a3839e32fd28e11cda60b3	FNCD	18
+\\xfb6f6467724b7cdc75a1977812d78b7467f69d77	FNCD	18
+\\xfc382c9f9f65a66bbd3b35e10e7ac5a7abf11c8f	FNCD	18
```
Are these different versions of the same token or what's going on here?

milkyklim replied:
Those are indeed different versions of the same token. Don't know how useful there are but there are in the list of tokens traded (at least once) on bancor.

milkyklim commented:
> Also, is the view_smart_tokens needed if all are added to erc20.tokens? Or perhaps you only added a subset to erc20.tokens?

@mewwts `erc20.tokens` contains normal tokens e.g. EOS or NMR while `view_smart_tokens` contains underlying bancor tokens, e.g. EOSBNT or NMRBNT. Does it make sense to you?

mewwts commented on schema/bancor/view_smart_tokens.sql:
```diff
@@ -0,0 +1,271 @@
+CREATE OR REPLACE VIEW bancor.view_smart_tokens (symbol, contract_address, decimals) AS VALUES
```
This should ideally match the order of `erc20.tokens` i.e. (contract_address, symbol, decimals)

milkyklim replied:
Sorry for that one, tricky to debug when can't really test the views. :) 
Committed order fix.

mewwts commented on schema/bancor/view_convert.sql:
```diff
@@ -0,0 +1,76 @@
+  (SELECT *
+   FROM erc20.tokens
+   UNION
+   SELECT *
```
See above comment, this union doesn't work unless the types are aligned.

mewwts commented:
Awesome! Thanks @milkyklim :+1:

mewwts commented:
![image](https://user-images.githubusercontent.com/1848670/82319512-4135bc80-99c1-11ea-9492-b6462efbe5c6.png)
Live on dune!
